File: packages/core/src/video-info/video-info.ts
import { execa } from "execa";

export class VideoInfo {
  private videoPath: string;
  constructor(videoPath: string) {
    this.videoPath = videoPath;
  }

  getFrameSize() {
    return new Promise<{ width: number; height: number }>(
      async (resolve, reject) => {
        try {
          const frameSizeProcess = execa("ffprobe", [
            "-v",
            "error",
            "-select_streams",
            "v:0",
            "-show_entries",
            "stream=width,height",
            "-of",
            "default=noprint_wrappers=1",
            `${this.videoPath}`,
          ]);
          frameSizeProcess.stdout?.on("data", (data: Buffer) => {
            const dataStr = data.toString();
            const widthMatch = dataStr.match(/width=(\d+)/);
            const heightMatch = dataStr.match(/height=(\d+)/);

            if (
              !widthMatch ||
              !widthMatch[1] ||
              !heightMatch ||
              !heightMatch[1]
            ) {
              return reject("no width and height found");
            }

            const widthStr = widthMatch[1];
            const heightStr = heightMatch[1];

            const width = parseInt(widthStr, 10);
            const height = parseInt(heightStr, 10);

            if (isNaN(width) || isNaN(height)) {
              throw "width and height not number";
            }

            return resolve({ width, height });
          });

          // ffprobe -v error -select_streams v:0 -show_entries stream=width,height -of default=noprint_wrappers=1 /Users/piyusharora/projects/via/assets/sample-videos/1-sec.mp4
        } catch (error) {
          reject(error);
        }
      }
    );
  }

  getFrameCount() {
    return new Promise<number>(async (resolve, reject) => {
      try {
        // ffprobe -v error -count_frames -select_streams v:0 -show_entries stream=nb_read_frames -of default=nokey=1:noprint_wrappers=1 /Users/piyusharora/projects/via/assets/sample-videos/1-sec.mp4
        const args = [
          "-v",
          "error",
          "-count_frames",
          "-select_streams",
          "v:0",
          "-show_entries",
          "stream=nb_read_frames",
          "-of",
          "default=nokey=1:noprint_wrappers=1",
          `${this.videoPath}`,
        ];

        const frameCountProcess = execa("ffprobe", args);
        frameCountProcess.stdout?.on("data", (data: Buffer) => {
          const dataStr = data.toString();
          const frameCount = parseInt(dataStr);
          if (isNaN(frameCount)) {
            reject("frame count not number");
          }

          resolve(frameCount);
        });
      } catch (error) {
        reject(error);
      }
    });
  }

  getFPS() {
    return new Promise<string>(async (resolve, reject) => {
      try {
        //ffprobe -v error -select_streams v:0 -show_entries stream=r_frame_rate -of default=noprint_wrappers=1:nokey=1 /Users/piyusharora/projects/via/assets/sample-videos/1-sec.mp4
        const args = [
          "-v",
          "error",
          "-select_streams",
          "v:0",
          "-show_entries",
          "stream=r_frame_rate",
          "-of",
          "default=noprint_wrappers=1:nokey=1",
          `${this.videoPath}`,
        ];

        const frameCountProcess = execa("ffprobe", args);
        frameCountProcess.stdout?.on("data", (data: Buffer) => {
          const fps = data.toString().replace(/\n/g, "");
          resolve(fps);
        });
      } catch (error) {
        reject(error);
      }
    });
  }
}

File: server-workspace/core/video-info/video-info.go
package videoinfo

import (
	"bytes"
	"os/exec"
	"strconv"
	"strings"
)

type GetFrameSizeOutput struct {
	Width  int
	Height int
}

func GetFrameSize(videoPath string) GetFrameSizeOutput {
	cmd := exec.Command("ffprobe", "-v", "error", "-select_streams", "v:0", "-show_entries", "stream=width,height", "-of", "default=noprint_wrappers=1", videoPath)

	var stdout, stderr bytes.Buffer

	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		println("ffprobe not found")
		println(stderr.String())
		panic(err)
	}

	cmdOut := strings.TrimSpace(stdout.String())
	tokens := strings.Split(cmdOut, "\n")
	widthStr := strings.Split(tokens[0], "width=")[1]
	heightStr := strings.Split(tokens[1], "height=")[1]

	width, widthErr := strconv.Atoi(widthStr)
	height, heightErr := strconv.Atoi(heightStr)

	if widthErr != nil || heightErr != nil {
		panic("width and height not found")
	}

	if width == 0 || height == 0 {
		panic("width and height not found")
	}

	output := GetFrameSizeOutput{
		Width:  width,
		Height: height,
	}

	return output
}
